components/servo/lib.rs

```diff
@@ -237,10 +244,46 @@ impl<Window> Browser<Window> where Window: WindowMethods + 'static {
         Browser {
             compositor: compositor,
             constellation_chan: constellation_chan,
+            embedder_receiver: embedder_receiver,
+        }
+    }
+
+    fn handle_window_event(&mut self, event: WindowEvent) {
+        match event {
+            WindowEvent::Reload => {
+                let top_level_browsing_context_id = match self.compositor.root_pipeline {
+                    Some(ref pipeline) => pipeline.top_level_browsing_context_id,
+                    None => return warn!("Window reload without root pipeline."),
+                };
+                let msg = ConstellationMsg::Reload(top_level_browsing_context_id);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!("Sending reload to constellation failed ({}).", e);
+                }
+            },
+            _ => {},
         }
     }
 
     pub fn handle_events(&mut self, events: Vec<WindowEvent>) -> bool {
+        while let Some(msg) = self.embedder_receiver.try_recv_embedder_msg() {
+            match (msg, self.compositor.shutdown_state) {
+                (EmbedderMsg::Status(message), ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.status(message);
+                },
+                (EmbedderMsg::ChangePageTitle(pipeline_id, title), ShutdownState::NotShuttingDown) => {
+                    let set_title = self.compositor.root_pipeline.as_ref().map_or(false, |root_pipeline| {
+                        root_pipeline.id == pipeline_id
+                    });
+                    if set_title {
+                        self.compositor.window.set_page_title(title);
+                    }
+                },
+                (_, _) => {},
+            }
+        }
+        for event in events.clone() {
```

Comment:
Good point. I would almost be tempted to think that all window events should be handled by Browser, which could call the appropriate method of compositor, or handle those differently(such as in this case, where the compositor isn't involved). Could this help support multiple compositors down the line? @paulrouget